src/components/tree/phyloTree/types.ts

```diff
@@ -0,0 +1,260 @@
+  scaleType: ScaleType | null
+  version: number
+  visibleLegendValues: string[]
+}
+
+export interface TemporalConfidence {
+  exists: boolean
+  display: boolean
+  on: boolean
+}
+
+export interface ReactTreeNode {
+  existingFilterState?: "active" | "inactive" | null
+  idx?: number
+  isBranch: boolean
+  name?: string
+  node?: PhyloNode
+  treeId?: string
+}
+
+export interface ReduxNode {
+  /** the index of the node in the nodes array. set so that we can access visibility / nodeColors if needed */
+  arrayIdx?: number
+
+  branch_attrs?: {
+    mutations?: {
+      [gene: string]: string[]
+    }
+    labels?: Record<string, any>
+  }
+  children?: ReduxNode[]
+  currentGt?: number
+  
+  /** see the number of subtending tips (alive or dead) */
+  fullTipCount?: number
+
+  hasChildren?: boolean
+  inView?: boolean
+  name: string
+  node_attrs?: {
+    div?: number
+    hidden?: "always" | "timetree" | "divtree"
+    num_date?: {
+      value: number
+    }
+  }
+  parent?: ReduxNode
+  parentInfo?: {
+    original?: ReduxNode
+  }
+  shell?: PhyloNode
+
+  /** the number of visible tips */
+  tipCount?: number
+
+  unexplodedChildren?: ReduxNode[]
+}
+
+/** [root idx tree1, root idx tree2] */
+export type Root = [number | undefined, number | undefined]
+
+/** Properties can be any property on PhyloNode but as an array for multiple nodes */
+export interface PropsForPhyloNodes {
+  branchStroke?: string[]
+  fill?: string[]
+  r?: number[]
+  tipStroke?: (string | undefined)[]
+  visibility?: Visibility[]
+}
+
+// Workaround for CSSStyleDeclaration having strokeWidth instead of stroke-width.
+// https://github.com/microsoft/TypeScript/issues/17827#issuecomment-861847433
+interface CSSStyleDeclarationWithHyphens extends CSSStyleDeclaration {
+  "stroke-width": string
+}
+
+// Workaround for CSSStyleDeclaration having generic string types.
+// Potentially related issue: https://github.com/Microsoft/TypeScript/issues/17827
+// TODO: consider extending existing interfaces such as SVGCircleElement for cx/cy/r
+interface CustomCSSStyleDeclaration {
+  cx?: number
+  cy?: number
+  d?: any
+  r?: number
+  "stroke-width"?: number
+  visibility?: Visibility
+
+  // TODO: This should be `string | number`, conditional on layout
+  x: any
+
+  // TODO: This should be `string | number`, conditional on layout
+  y: any
+}
+
+// Apply fixes from above
+export type SVGProperty = Extract<keyof CSSStyleDeclarationWithHyphens | keyof CustomCSSStyleDeclaration, string>
+
+export interface PhyloNode extends Partial<Omit<CSSStyleDeclarationWithHyphens, keyof CustomCSSStyleDeclaration>>, CustomCSSStyleDeclaration {
+  angle?: number
+  branch?: [string, string]
+  branchStroke?: string
+  conf?: [number, number]
+
+  /** SVG path */
+  confLine?: string
+
+  crossDepth?: number
+  depth?: number
+  displayOrder?: number
+  displayOrderRange?: number[]
+  fill?: string
+  inView: boolean
+  n: ReduxNode
+  pDepth?: number
+  px?: number
```

Comment:
This is what I was trying to do with b37a77c9555b11907c6698e6a1d07ab94533d00f. I think it has to start at the `PhyloTree` level since that is where layout is set and checked. I've made a second attempt using [mapped types](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html): 9da17b783f1e1d4d4c0fc05b33a639dc883c09b8

Reply:
Oh cool! Shall we add those commits as a separate PR (I realise they're not part of this one) and we can continue discussion there? I think it'd be a good direction, but unsure of the difficulty / priority.

Reply:
I have them in this PR currently (the commit links go stale upon force-push) but yeah I think it'd be better to consider separately. Added a post-merge task:

> - Post-merge: consider using mapped types in a new PR